### PR TITLE
refactor(tree2): Use cursor with adapter to implement mapping from proxy domain to "edit tree" domain

### PR DESCRIPTION
**Pull request by Josmithr**
This PR strictly updates how trees are mapped when inserting new content. Updating the handling of initialTree data will be done in a subsequent PR.

AB#6059

### CODE REVIEW

File: experimental/dds/tree2/.prettierignore [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Generated by npm
```
Comment (Josmithr) [inline experimental/dds/tree2/.prettierignore:1]: Extracted these changes into their own PR: #18280

File: experimental/dds/tree2/src/shared-tree/schematizedTree.ts [hunk 2]
```diff
@@ -220,6 +220,7 @@ export interface TreeContent<TRoot extends TreeFieldSchema = TreeFieldSchema>
 	 * Default tree content to initialize the tree with iff the tree is uninitialized
 	 * (meaning it does not even have any schema set at all).
 	 */
+	// TODO: change to proxy types?
```
Comment (Josmithr) [inline experimental/dds/tree2/src/shared-tree/schematizedTree.ts:223]: Will be done in a subsequent PR
```suggestion
```

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/toEditTree.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+				return [];
+			}
+
+			if (node === null) {
+				return [];
+			}
+
+			if (Array.isArray(node)) {
+				// TODO: verify this
+				return key === EmptyKey || key === rootFieldKey ? node : [];
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/toEditTree.ts:130]: @noencke @CraigMacomber does this look right to you? The json cursor implementation only looks at `EmptyKey`, but we have at least some tests with an array as the root field, and they were failing without us also accepting the `rootFieldKey`.

File: experimental/dds/tree2/src/domains/json/jsonCursor.ts [hunk 4]
```diff
@@ -21,7 +21,7 @@ const adapter: CursorAdapter<JsonCompatible> = {
 	value: (node: JsonCompatible) =>
 		node !== null && typeof node === "object"
 			? undefined // arrays and objects have no defined value
-			: node, // null, boolean, numbers, and strings are their own value
+			: node, // null, boolean, numbers, strings, and Fluid Handles are their own value
```
Comment (Josmithr) [inline experimental/dds/tree2/src/domains/json/jsonCursor.ts:24]:
```suggestion
			: node, // null, boolean, numbers, and strings are their own values
```

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+): CursorAdapter<TNode> {
+	return {
+		value: (node: TNode) => {
+			return isPrimitiveValue(node) || node === null || isFluidHandle(node)
+				? (node as Value)
+				: undefined;
+		},
+		type: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error("undefined node");
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:48]: Is it possible to hit this? If the schema is `{ foo?: number }` and the user provides `{}`, then certainly this won't be hit. But what if they provide `{ foo: undefined }`? Do we handle that case? 

If we do, I would expect it to be handled in a layer above - whatever code is passing stuff to the cursor adaptor should filter out explicit undefineds. If that is the case, then this should probably be an `assert` rather than an Error (asserts are typically better for internal invariants being broken, whereas Errors are better for things a user might cause).

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+			}
+			if (node === null) {
+				return leaf.null.name;
+			}
+
+			if (isFluidHandle(node)) {
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:54]: Can we move this into the `default` part of the switch statement? It's slower than checking `number`, `string`, or `boolean`, so it's a little micro-op to have it be checked after those. I don't want to _prematurely_ micro-op, but in this case it doesn't have any code complexity cost to move it there.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+					return possibleTypes[0];
+				}
+			}
+		},
+		keysFromNode: (node: TNode) => {
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:88]: This doesn't match the check in `getPossibleTypes`, since that one uses `isArrayLike`. We probably want to get rid of `isArrayLike`, but until then, this should probably be in sync with that (or have a TODO).

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+						return node.length === 0 ? [] : [EmptyKey];
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:89]: Why do we not have the EmptyKey field for empty arrays? Is this an optimization?

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+					} else if (node instanceof Map) {
+						const unfilteredKeys = Array.from(node.keys()) as FieldKey[];
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:91]: It's probably more efficient to filter the `entries()` of the map, and then `map` that to only produce the keys. Then no map lookups are required.

Same thing for the object case below.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+				// TODO: verify this
+				return key === EmptyKey || key === rootFieldKey ? node : [];
+			}
+
+			if (node instanceof Map) {
+				return (node as Map<FieldKey, TNode[]>).get(key) ?? [];
+			}
+
+			if (Object.prototype.hasOwnProperty.call(node, key)) {
+				const field = (node as Record<FieldKey, TNode>)[key];
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:138]: Do we need to [exclude explicit undefineds](https://www.typescriptlang.org/play?#code/MYewdgziA2CmB00QHMAUB5ARgK1sALvAA4BOI+5AnkQgBYCGE6A7mAApk0n6XzD3RoqAN4ACAB4AuUQFcwAE1gAzAJZhY80QF8ANKIBE4-QEpjAbiA)?

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 5 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:48]: I was hitting this while my code had bugs. I don't remember the exact scenario where it was happening though :/

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 10 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:138]: Oh, yeah we probably do.

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:138]: No. We filter out keys with undefined values in `keysFromNode`. So presumably this doesn't get called for those keys. If it does, I believe it's valid to just return undefined.

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:138]: Yeah, you're probably right. We already filter out keys with undefined values in `keysFromNode`, so presumably this doesn't get called. But in the event that it was called, this would be incorrect. Will update.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 5 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts]: I wasn't able to repro again, but note that the ProxyNode type includes `unknown` which can allows for `undefined`. We don't expect `undefined` to be provided, so I think explicit error boundaries here is appropriate.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 7 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:95]: I don't think we need that here, since we aren't supported input formats that aren't arrays in the proxy layer, right?

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 8 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:98]: Per our investigation, this appears to be:
a. Just part of the contract for lists
b. Consumers look for empty lists to signal no field keys, so returning an empty array will break some checks.

This contract seems ill specified IMO, but appears to be important.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 9 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:100]: Map lookups amortize to O(1), and copying all of the entries into an array consumes more space than just an array of keys. So, I think this is the more optimal approach.

If this was returning an iterable, the approach you suggested would probably fit better, but we actually have to return an array here.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 7 above)
Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:95]: Yea, you're right. I guess it will resolve itself when we copy over and adjust all the validation code.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ * @param mode - The mode of cursor creation.
+ * Determines whether the cursor is initialized in "node" mode or "field" mode.
+ */
+export function cursorFromProxyTree<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	node: TNode,
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+	mode: "node" | "field" = "node",
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:176]: Is it appropriate to use `CursorLocationType` here, or is that a weird import?

**Review by noencke — COMMENTED** [review]
Tests look good!

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts (hunk 11 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts:176]: Wasn't aware that was a thing. I'll give it a shot.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+ * Consumers should use {@link cursorFromProxyTree} instead.
+ */
+export function createProxyTreeAdapter<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+): CursorAdapter<TNode> {
+	return {
+		value: (node: TNode) => {
+			return isPrimitiveValue(node) || node === null || isFluidHandle(node)
+				? (node as Value)
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:50]: I'm surprised this cast is needed. Can it be removed? If now, why?

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:50]: It was at one point in time, but the code has been shuffled significantly since. I'll verify and remove if possible,

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+				case "string":
+					return leaf.string.name;
+				case "boolean":
+					return leaf.boolean.name;
+				default: {
+					if (isFluidHandle(node)) {
+						return leaf.handle.name;
+					}
+
+					const possibleTypes = getPossibleTypes(
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:75]: At this point you have a ProxyNode right? Don't you just want Tree.schema(node).name here?

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+			}
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
+						// A list with no children is considered to not have any fields.
+						// If it has children, return the special symbol indicating a list.
+						return node.length === 0 ? [] : [EmptyKey];
+					} else if (node instanceof Map) {
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:106]: Are TreeMapNodes nodes actually instances of map?

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts (hunk 13 above)
Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:75]: `node` is a `ProxyNode<"javaScript">`, not a `ProxyNode<"sharedTree">`. Only the latter can be used in such a way. The stuff being processed here is essentially POJO.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts (hunk 14 above)
Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:106]: No, but the object that you use in inserted content to specify a new Map is. These aren't `TreeMapNodes`, they're just POJOs with Arrays and Maps in them sometimes.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxies.ts [hunk 15]
```diff
@@ -614,12 +646,23 @@ const mapStaticDispatchMap: PropertyDescriptorMap = {
...
 			const { content, hydrateProxies } = extractFactoryContent(
 				value as FlexibleFieldContent<MapFieldSchema>,
 			);
+			const mappedContent =
+				content === undefined
+					? undefined
+					: cursorFromProxyTree(
+							content,
+							node.context,
+							node.schema.mapFields?.types ??
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxies.ts:660]: After the my latest PR that you reviewed, `mapFields` should be guaranteed to be present here now (since `node.schema` is now strongly typed as a MapNodeSchema)

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts (hunk 14 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:106]: `SharedTreeMap` (the proxy type) is typed as a Map, yes. Assuming that's what you're referring to.

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:106]: I believe the intention is to allow other forms of input in the future, but for now we strictly take TypeScript Maps.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+ * Creates a {@link CursorAdapter} capable of mapping proxy-domain trees to the editable tree domain.
+ *
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ *
+ * @privateRemarks
+ * This is exported solely for testing purposes.
+ * Consumers should use {@link cursorFromProxyTree} instead.
+ */
+export function createProxyTreeAdapter<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:43]: I don't think making this generic adds any value. Sadly I don't think the proxy API provides any generic tree reading friendly types, and its not covariant over schema, so there isn't really a well-formed equivalent to FlexTreeNode.

Anyway, this generic setup is also wrong, since the adapter below assumes all children of a node are the same type as the node.

Something like `type TreeEntity = TreeValue | { readonly [P in string]?: TreeEntity } | ReadonlyMap<string|TreeEntity>` is the type you want I guess?

The lack of read only APIs preventing covariance over schema, and lack of type safety for proxy objects vs other objects makes thus such a mess.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+			}
+
+			if (node instanceof Map) {
+				const value = (node as Map<FieldKey, TNode>).get(key);
+				return value === undefined ? [] : [value];
+			}
+
+			if (Object.prototype.hasOwnProperty.call(node, key)) {
+				const field = (node as Record<FieldKey, TNode>)[key];
+				return field === undefined ? [] : [field];
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:161]: Daniel is supposed to fix that we report proxies as having keys when the field is undefined, so this check should really be an assert (or t lest have a TODO to become one when thats fixed)

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+	mode: CursorLocationType = CursorLocationType.Nodes,
+): CursorWithNode<TNode> {
+	const adapter = createProxyTreeAdapter<TNode>(context, typeSet);
+	switch (mode) {
+		case CursorLocationType.Nodes:
+			return stackTreeNodeCursor(adapter, node);
+		case CursorLocationType.Fields:
+			return stackTreeFieldCursor(adapter, node);
+		default:
+			throw new Error(`Unrecognized CursorLocationType: "${mode}."`);
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:191]: use unreachableCase for defaults when matches should be exhastive to get compile time checking.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts (hunk 13 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:75]: I'll give that a shot. I tried using that at one point and it didn't work, but that was early on.

File: experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts (hunk 17 above)
Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts:161]: But `node` is not a proxy. It is just a POJO.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/toEditTree.ts (hunk 3 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/toEditTree.ts]: the simple tree API does not allow sequence fields in any place other than as the EmptyKey field of a list node. Thus it does not support sequences at the root. Infact it doesn't really support non-required roots very well as currently there is no way to edit or really access the root field.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/toEditTree.ts]: Also, I don't think using Array.isArray here is a good idea. The input here is supposed to be able to be a proxy node tree thing right? List nodes return false for Array.isArray right?

Instead shouldn't you check the schema of the node, and see if its a list schema?
